### ISSUE TITLE
Android build

### ISSUE THREAD

**Issue by selaux** [ja2-stracciatella/ja2-stracciatella#228]
Good news everyone :tada:! I compiled the `sdl2` branch for Android using the example project provided by SDL2. This was surprisingly easy, barely any code changes needed to be done and the result is a working apk, which could in theory be played. There are still issues with font rendering though, that were fixed in the SDL1.2 port.

You can try [SDLActivity-debug.apk.zip](https://github.com/ja2-stracciatella/ja2-stracciatella/files/149328/SDLActivity-debug.apk.zip) on your phone. For simplicity I hardcoded all data paths to `"/storage/sdcard1/ja2-data"` which will hopefully be present on most phones (with sd cards). Your `data` and `externalized` directories need to be copied there.

Things learned from this port:
- We should update to boost 1.60.0, because it fixes compilation of boost for Android.
- Dirty rectangle screen updating seems to be broken on Android (at least on my i9300)
- The touch input needs updating :wink:
- Is the error-catching around the `main` function necessary? This hid away a lot of errors when building ja2 as a library.

I can upload some code for the Android wrapper prototype if you want. This might have to be done in a different repository though, or should we keep everything in here.

**Comment by lynxlynxlynx** [comment]
I'm in contact with Omnigraphical, the author of the first port. Heh, I told him/her an hour ago that work on this is not a priority for the rest of us before the SDL2 port is complete. I guess I have to bite my tongue now. :)

Hardcoding the data path is a very bad thing(tm) in the long run. Android systems are a nightmare in this regard.

I think we should keep everything in the same repo and codebase. That's the whole point of convergence and better maintainability. Sure some #ifdef ANDROID will be needed, but that's no different than for the rest of the platforms.
However, please don't merge this work with your sld2 branch, at least not until it gets merged.

**Comment by selaux** [comment]
I was annoyed by the one-off error in the sdl2 branch, so I took some time off :smile:. I wont merge anything into the sdl2 branch, this is just to prove that its easier than with sdl1.

Putting everything into this repo will include some android specific build stuff and `*.java` files though, but maybe we can confine this into a subdirectory.

**Comment by derSteFfi** [comment]
i'm currently experimenting with externalizing some of the libs to my host version instead of the delivered one. Libboost (1.60) works fine for me, I'm about to put together a nice patch and PR it. Will be here about tomorrow afternoon (in Europe). Of course, like for libSDL, with a switch to fall back on the version delivered in this repo.

**Comment by selaux** [comment]
Sounds good. I also tried compiling and running with boost 1.60.0 outside of this branch (I tested building with cmake instead of the Makefile). Worked fine.

**Comment by selaux** [comment]
Seems `wchar_t` support is kind of broken on Android :frowning:

**Comment by lynxlynxlynx** [comment]
Yes, that was the bulk of the work that needed to be done. Omni mentioned that it perhaps won't be an issue anymore with a more recent toolchain.

**Comment by selaux** [comment]
I dont know if that was the `wchar_t` stuff or anything else, but I switched around std libraries (now using shared gnu libraries) and could resolve it to print strings correctly without having to do any code changes. Now there aren't any apparent errors anymore. I have not playtested it thoroughly though, since the right-click functionality is missing :no_entry:. I attached another build that works fine on the devices I have (armeabi and arm64-v8a) other abis are included as well. Ini editing is no longer necessary, the config is now stored in `SDL_GetPrefPath`, but the data and externalized dirs are still hardcoded to the external sd card.

[SDLActivity-debug.apk.zip](https://github.com/ja2-stracciatella/ja2-stracciatella/files/149328/SDLActivity-debug.apk.zip)

**Comment by lynxlynxlynx** [comment]
That's great news indeed. I hope it'll be fine with non-English too. Is right-click critical for anything else but aiming?

**Comment by griffar** [comment]
hi, could you build a version using a data path pointing to /sdcard/ja2-data please?

thx

**Comment by selaux** [comment]
@griffar: Here is the newest version it stores its configuration as well as its data files in `/Android/data/org.ja2stracciatella.ja2/files` on your internal storage. Start the app once then copy `data` and `externalized` to this directory.

We get that from `SDL_AndroidGetExternalStoragePath`, but I think we need to mix it with `SDL_GetPrefPath` because it gets removed when you delete the app (and so are all the savegames).

[SDLActivity-debug.apk.zip](https://github.com/ja2-stracciatella/ja2-stracciatella/files/149440/SDLActivity-debug.apk.zip)

**Comment by selaux** [comment]
I just finished the first sector and have to say that with the current input system the game is not fun to play  :grin:. Right click is important for a lot of things (e.g. opening the inventory), scrolling is bugged. But for an early port in runs fine.

**Comment by Srandista** [comment]
I was going to try latest debug version on my Nexus 7 with Android 6.0.1, but I received following error:

SDL Error
An error occurred while trying to start the application. Please try again and/or reinstall.

Error: dlopen failed: couldn't reserve 1083441152 bytes of address space for "/data/app/org.ja2stracciatella.ja2-1/lib/arm/libmain.so"

Can something be done about it?

**Comment by Manuel-K** [comment]
That does sound like the gfkeystate problem that was fixed in #523.

@selaux Can you build a new android package?

**Comment by selaux** [comment]
Unfortunately I can't that easily. Basically those were just test builds i did with build system code that I don't have anymore. I tried another attempt at a proper Android build system a few months ago, as Android Studio now has cmake build support, but then there were issues with packaging the imported shared libraries that we use (e.g. SDL2 or libstracciatella). So doing a rebuild would currently require me to code the whole build stuff again...

**Comment by Srandista** [comment]
I don't need some test version now. I was just thinking about helping you guys with testing, if it's needed.

I just wanna know, for version 0.16/1.00, is there still plan to release Android version as well?

**Comment by lynxlynxlynx** [comment]
I was fixing another android port and noticed this explanation about why dirty rect updates don't work for SDL on Android:

>  A note regarding the use of the "dirty rectangles" rendering technique
> 
> If your app uses a variation of the "dirty rectangles" rendering technique, where you only update a portion of the screen on each frame, you may notice a variety of visual glitches on Android, that are not present on other platforms. 
> This is caused by SDL's use of EGL as the support system to handle OpenGL ES/ES2  contexts, in particular the use of the eglSwapBuffers function. As stated in the documentation for the function "The contents of ancillary buffers are always undefined after calling eglSwapBuffers".
> Setting the EGL_SWAP_BEHAVIOR attribute of the surface to EGL_BUFFER_PRESERVED is not possible for SDL as it requires EGL 1.4, available only on the API level 17+, so the only workaround available on this platform is to redraw the entire screen each frame.
> 
> Reference: http://www.khronos.org/registry/egl/specs/EGLTechNote0001.html

**Comment by lynxlynxlynx** [comment]
And RE the rightclick problem: just came across our WITH_MAEMO build switch, which means "Build with right click mapped to F4 (menu button)". For android we could just map to one of the physical keys (easily made configurable) or use a gesture (in the multitouch sense).

**Comment by sidav** [comment]
How can I try to build JA2-Stracciatella for Android?

**Comment by selaux** [comment]
Currently it is not possible as it requires some custom build setup that is not yet integrated into the repository. I can attach the files to this ticket if you want, but I'm not even sure if they still build...

**Comment by selaux** [comment]
I think we can close this, as the build now happens automatically for nightlies. https://github.com/ja2-stracciatella/ja2-stracciatella/issues/1250 and https://github.com/ja2-stracciatella/ja2-stracciatella/issues/1248 are the follow up items for "proper" Android support (including touch).
